astropy/utils/masked/function_helpers.py

```diff
@@ -1433,6 +1433,145 @@ def setdiff1d(ar1, ar2, assume_unique=False):
     return ar1[np.isin(ar1, ar2, assume_unique=True, invert=True)], None, None
 
 
+@dispatched_function
+def einsum(*operands, out=None, **kwargs):
+    from astropy.utils.masked import Masked, MaskedNDArray
+
+    # input validation
+    subscripts, *operands = operands
+
+    if not isinstance(subscripts, str):
+        raise ValueError('only "subscripts" string mode supported for einsum.')
+
+    kwargs_with_out = kwargs.copy()
+    if out is not None:
+        if not isinstance(out, MaskedNDArray):
+            raise NotImplementedError
+
+        else:
+            kwargs_with_out["out"] = out.unmasked.view(np.ndarray)
+
+    # compute the unmasked result
+    # doing this first ensures that subscripts are validated by numpy
+
+    def as_masked_ndarray(input):
+        if hasattr(input, "mask"):
+            if hasattr(input.mask, "shape"):
+                return input
+            else:
+                mask = np.full_like(input.unmasked, fill_value=input.mask, dtype=bool)
+                return Masked(input.unmasked, mask=mask)
+        else:
+            mask = np.full_like(input, fill_value=False, dtype=bool)
+            return Masked(input, mask=mask)
+
+    mas = tuple(as_masked_ndarray(op) for op in operands)
+    arrays = tuple(ma.unmasked for ma in mas)
+    retv = np.einsum(subscripts, *arrays, **kwargs_with_out)
+
+    # now combine masks
+    # this complex operation is decomposed as
+    # 1) Compute intermediate masks
+    #   1.1) select diagonals (e.g. 'ii')
+    #   1.2) contract mask along summation axes
+    # 2) combine intermediate masks
+
+    masks = tuple(ma.mask for ma in mas)
+
+    from collections import Counter
```

Comment:
Are you sure this whole sequence is necessary? If we stick with that any masked element propagates correctly, then can we not just use the same trick as for matmul in `__array_ufunc__`: https://github.com/astropy/astropy/blob/0ea3682a4586934a40d30d706010fda0314e6b18/astropy/utils/masked/core.py#L780-L794

(Note that it did take me quite a while to realize this might work...)